navigation/pose_action_server/src/pose_action_server_ros.cpp

```diff
@@ -0,0 +1,148 @@
+    const geometry_msgs::msg::PoseStamped::ConstSharedPtr pose_msg) {
+    if (!is_executing_action_)
+        return;
+
+    pose_queue_.push_back(*pose_msg);
+
+    geometry_msgs::msg::PoseStamped feedback_msg = *pose_msg;
+    auto feedback =
+        std::make_shared<vortex_msgs::action::FilteredPose::Feedback>();
+    feedback->current_pose = feedback_msg;
+    active_goal_handle_->publish_feedback(feedback);
+
+    if (pose_queue_.size() == num_measurements_) {
+        geometry_msgs::msg::PoseStamped filtered_pose;
+        filtered_pose.header = pose_msg->header;
+
+        std::vector<double> x_positions;
+        std::vector<double> y_positions;
+        std::vector<double> z_positions;
+        std::vector<double> yaw_angles;
+
+        for (const auto& pose : pose_queue_) {
+            x_positions.push_back(pose.pose.position.x);
+            y_positions.push_back(pose.pose.position.y);
+            z_positions.push_back(pose.pose.position.z);
+
+            double qz = pose.pose.orientation.z;
+            double qw = pose.pose.orientation.w;
+            double yaw = 2.0 * std::atan2(qz, qw);
+            yaw_angles.push_back(yaw);
+        }
+
+        std::sort(x_positions.begin(), x_positions.end());
+        std::sort(y_positions.begin(), y_positions.end());
+        std::sort(z_positions.begin(), z_positions.end());
+
+        size_t mid = num_measurements_ / 2;
+
+        if (num_measurements_ % 2 == 0) {
+            filtered_pose.pose.position.x =
+                (x_positions[mid - 1] + x_positions[mid]) / 2.0;
+            filtered_pose.pose.position.y =
+                (y_positions[mid - 1] + y_positions[mid]) / 2.0;
+            filtered_pose.pose.position.z =
+                (z_positions[mid - 1] + z_positions[mid]) / 2.0;
+        } else {
+            filtered_pose.pose.position.x = x_positions[mid];
+            filtered_pose.pose.position.y = y_positions[mid];
+            filtered_pose.pose.position.z = z_positions[mid];
+        }
+
+        std::vector<double> unwrapped_yaw = yaw_angles;
+        for (size_t i = 1; i < unwrapped_yaw.size(); ++i) {
+            double diff = unwrapped_yaw[i] - unwrapped_yaw[i - 1];
+            if (diff > M_PI) {
+                unwrapped_yaw[i] -= 2.0 * M_PI;
+            } else if (diff < -M_PI) {
+                unwrapped_yaw[i] += 2.0 * M_PI;
+            }
+        }
+
+        std::sort(unwrapped_yaw.begin(), unwrapped_yaw.end());
+        double median_yaw = unwrapped_yaw[mid];
+        if (num_measurements_ % 2 == 0) {
+            median_yaw = (unwrapped_yaw[mid - 1] + unwrapped_yaw[mid]) / 2.0;
+        } else {
+            median_yaw = unwrapped_yaw[mid];
+        }
+
+        double filtered_yaw = median_yaw;
+        filtered_pose.pose.orientation.x = 0.0;
+        filtered_pose.pose.orientation.y = 0.0;
+        filtered_pose.pose.orientation.z = std::sin(filtered_yaw / 2.0);
+        filtered_pose.pose.orientation.w = std::cos(filtered_yaw / 2.0);
```

Comment:
Just a sanity check here, what's the reasoning for picking the median instead of the mean? I wouldn't expect an EKF (assuming that's what's used from the topic naming) to produce nasty outliers f.ex.

(If you used the mean there's a fairly well known paper detailing a linear-time (in N quats) quaternion mean that doesnt involve transforming your quats to euler angles [here](https://www.acsu.buffalo.edu/%7Ejohnc/ave_quat07.pdf), which with all equal weights becomes even more straight forward)

Reply:
That's fair, It would be interesting to see the full system behavior with averaging across 1, 10, 100 etc etc measurements. Since you are fusing with a bayes filter you have (under gaussian/linear assumptions and no nasty linearization errors) the "best possible" estimate already - at least in the information sense